crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,336 @@
+                newer.fields(($(cb::$field),*))
+            }
+        }
+
+        // Construct the basic query first to filter by owner, not deleted and newest rows.
+        let mut query = candidates
+            .select(candidates!(
+                object_id,
+                cp_sequence_number,
+                coin_balance_bucket
+            ))
+            .left_join(
+                newer.on(candidates!(object_id)
+                    .eq(newer!(object_id))
+                    .and(candidates!(cp_sequence_number).lt(newer!(cp_sequence_number)))),
+            )
+            .filter(newer!(object_id).is_null())
+            .filter(candidates!(owner_id).eq(owner.to_vec()))
+            .filter(candidates!(coin_balance_bucket).is_not_null())
+            .into_boxed();
+
+        // If the coin type is specified, we filter by it.
+        if let Some(coin_type_tag) = coin_type_tag {
+            let serialized_coin_type =
+                bcs::to_bytes(&coin_type_tag).context("Failed to serialize coin type tag")?;
+            query = query.filter(candidates!(coin_type).eq(serialized_coin_type));
+        }
+
+        // If the cursor is specified, we filter by it.
+        if let Some(Cursor(BalanceCursor {
+            object_id,
+            cp_sequence_number,
+            coin_balance_bucket,
+        })) = page.cursor
+        {
+            query = query
+                .filter(candidates!(cp_sequence_number).le(cp_sequence_number as i64))
+                .filter(
+                    // Since the result is ordered by coin_balance_bucket followed by object_id,
+                    // we need to filter by coin_balance_bucket first, then if balance bucket is the same, object_id.
+                    candidates!(coin_balance_bucket)
+                        .lt(coin_balance_bucket)
+                        .or(candidates!(coin_balance_bucket)
+                            .eq(coin_balance_bucket)
+                            .and(candidates!(object_id).gt(object_id.to_vec()))),
+                );
+        }
+
+        // Finally we order by coin_balance_bucket and break ties by object_id.
+        query = query
+            .order_by(candidates!(coin_balance_bucket).desc())
+            .then_order_by(candidates!(object_id).asc());
+
+        let mut buckets: Vec<IdCpBalance> =
+            conn.results(query).await.context("Failed to query coins")?;
+
+        // Now gather pagination info.
+        let has_next_page = buckets.len() > limit as usize;
+        if has_next_page {
+            buckets.truncate(limit as usize);
+        }
+
+        let next_cursor = buckets
+            .last()
+            .map(|bucket| {
+                let object_id =
+                    ObjectID::from_bytes(&bucket.object_id).context("Failed to parse object id")?;
+                let balance_bucket = bucket
+                    .coin_balance_bucket
+                    .ok_or_else(|| anyhow::anyhow!("Coin balance bucket is null"))?;
```

Comment:
I don't think I'm fully grasping how error handling should properly work here 😅
`.context` turns the error into an `anyhow::Error` so here I use `anyhow::anyhow`.
Would like @amnn 's advice on if this is the right way.